nanonameserver.py

```diff
@@ -0,0 +1,194 @@
+
+    If a port is not specified, random ports will be chosen.
+
+    Applications should subclass the server and override the handle()
+    method to determine how the server responds to queries.  The
+    default behavior is to refuse everything.
+
+    If use_thread is set to False in the constructor, then the
+    server's main() method can be used directly in a trio nursery,
+    allowing the server's cancellation to be managed in the Trio way.
+    In this case, no thread creation ever happens even though Server
+    is a subclass of thread, because the start() method is never
+    called.
+    """
+
+    def __init__(self, address='127.0.0.1', port=0, enable_udp=True,
+                 enable_tcp=True, use_thread=True):
+        super().__init__()
+        self.address = address
+        self.port = port
+        self.enable_udp = enable_udp
+        self.enable_tcp = enable_tcp
+        self.use_thread = use_thread
+        self.left = None
+        self.right = None
+        self.udp = None
+        self.udp_address = None
+        self.tcp = None
+        self.tcp_address = None
+
+    def __enter__(self):
+        (self.left, self.right) = socket.socketpair()
+        # We're making the UDP socket now so it can be sent to by the
+        # caller immediately (i.e. no race with the listener starting
+        # in the thread).
+        if self.enable_udp:
+            self.udp = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, 0)
+            self.udp.bind((self.address, self.port))
+            self.udp_address = self.udp.getsockname()
+        if self.enable_tcp:
+            self.tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
+            self.tcp.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            self.tcp.bind((self.address, self.port))
+            self.tcp.listen()
+            self.tcp_address = self.udp.getsockname()
+        if self.use_thread:
+            self.start()
+        return self
+
+    def __exit__(self, ex_ty, ex_va, ex_tr):
+        if self.left:
+            self.left.close()
+        if self.use_thread and self.is_alive():
+            self.join()
+        if self.right:
+            self.right.close()
+        if self.udp:
+            self.udp.close()
+        if self.tcp:
+            self.tcp.close()
+
+    async def wait_for_input_or_eof(self):
+        #
+        # This trio task just waits for input on the right half of the
+        # socketpair (the left half is owned by the context manager
+        # returned by launch).  As soon as something is read, or the
+        # socket returns EOF, EOFError is raised, causing a the
+        # nursery to cancel all other nursery tasks, in particular the
+        # listeners.
+        #
+        try:
+            with trio.socket.from_stdlib_socket(self.right) as sock:
+                self.right = None  # we own cleanup
+                await sock.recv(1)
+        finally:
+            raise EOFError
+
+    def handle(self, message):
```

Comment:
Should the client address and/or protocol be passed into `handle()`? Either independently, or as part of a query context that could be extended later.